File: Palace/Catalog/TPPOpenSearchDescription.m
#import "TPPAsync.h"
#import "TPPSession.h"
#import "TPPXML.h"
#import "NSString+TPPStringAdditions.h"
#import "Palace-Swift.h"

#import "TPPOpenSearchDescription.h"

@interface TPPOpenSearchDescription ()

@property (nonatomic) NSString *humanReadableDescription;
@property (nonatomic) NSString *OPDSURLTemplate;
@property (nonatomic) NSArray *books;

@end

@implementation TPPOpenSearchDescription

+ (void)withURL:(NSURL *const)URL
shouldResetCache:(BOOL)shouldResetCache
completionHandler:(void (^)(TPPOpenSearchDescription *))handler
{
  if(!handler) {
    @throw NSInvalidArgumentException;
  }
  
  [[TPPSession sharedSession]
   withURL:URL
   shouldResetCache:shouldResetCache
   completionHandler:^(NSData *const data, __unused NSURLResponse *response, __unused NSError *error) {
     if(!data) {
       TPPLOG(@"Failed to retrieve data.");
       TPPAsyncDispatch(^{handler(nil);});
       return;
     }
     
     TPPXML *const XML = [TPPXML XMLWithData:data];
//     NSString *datcat = [[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] substringToIndex:100];
//     NSDictionary *errData = @{@"data": [NSString stringWithFormat:@"%@...", datcat]};
     if(!XML) {
       TPPLOG(@"Failed to parse data as XML.");
       TPPAsyncDispatch(^{handler(nil);});
       return;
     }
     
     TPPOpenSearchDescription *const description =
       [[TPPOpenSearchDescription alloc] initWithXML:XML];
     
     if(!description) {
       TPPLOG(@"Failed to interpret XML as OpenSearch description document.");
       TPPAsyncDispatch(^{handler(nil);});
       return;
     }
     
     TPPAsyncDispatch(^{handler(description);});
   }];
}

- (instancetype)initWithXML:(TPPXML *const)OSDXML
{
  self = [super init];
  if(!self) return nil;
  
  self.humanReadableDescription = [OSDXML firstChildWithName:@"Description"].value;
  
  if(!self.humanReadableDescription) {
    TPPLOG(@"Missing required description element.");
    return nil;
  }
  
  for(TPPXML *const UrlXML in [OSDXML childrenWithName:@"Url"]) {
    NSString *const type = UrlXML.attributes[@"type"];
    if(type && [type rangeOfString:@"opds-catalog"].location != NSNotFound) {
      self.OPDSURLTemplate = UrlXML.attributes[@"template"];
      break;
    }
  }
  
  if(!self.OPDSURLTemplate) {
    TPPLOG(@"Missing expected OPDS URL.");
    return nil;
  }
  
  return self;
}

- (instancetype)initWithTitle:(NSString *)title books:(NSArray *)books
{
  self = [super init];
  if(!self) return nil;
  self.books = books;
  self.humanReadableDescription = title;
  return self;
}

- (NSURL *)OPDSURLForSearchingString:(NSString *)searchString
{
  NSString *urlStr = [self.OPDSURLTemplate
                      stringByReplacingOccurrencesOfString:@"{searchTerms}"
                      withString:[searchString stringURLEncodedAsQueryParamValue]];
  return [NSURL URLWithString:urlStr];
}

@end

File: Palace/Catalog/TPPOpenSearchDescription.swift
import Foundation

@objc class TPPOpenSearchDescription: NSObject {

  @objc private(set) var humanReadableDescription: String?
  @objc var opdsURLTemplate: String?
  @objc private(set) var books: [Any]?

  private override init() {
    super.init()
  }

  @objc static func withURL(
    _ url: URL,
    shouldResetCache: Bool,
    completionHandler handler: @escaping (TPPOpenSearchDescription?) -> Void
  ) {
    TPPSession.sharedSession.withURL(url, shouldResetCache: shouldResetCache) { data, _, _ in
      guard let data = data else {
        Log.log("Failed to retrieve data.")
        TPPAsyncDispatch { handler(nil) }
        return
      }

      guard let xml = TPPXML.xml(withData: data) else {
        Log.log("Failed to parse data as XML.")
        TPPAsyncDispatch { handler(nil) }
        return
      }

      guard let description = TPPOpenSearchDescription(xml: xml) else {
        Log.log("Failed to interpret XML as OpenSearch description document.")
        TPPAsyncDispatch { handler(nil) }
        return
      }

      TPPAsyncDispatch { handler(description) }
    }
  }

  @objc init?(xml osdXML: TPPXML) {
    super.init()

    humanReadableDescription = osdXML.firstChild(withName: "Description")?.value
    guard humanReadableDescription != nil else {
      Log.log("Missing required description element.")
      return nil
    }

    for urlXML in osdXML.childrenWithName("Url") {
      if let type = (urlXML.attributes as? [String: String])?["type"],
         type.contains("opds-catalog") {
        opdsURLTemplate = (urlXML.attributes as? [String: String])?["template"]
        break
      }
    }

    guard opdsURLTemplate != nil else {
      Log.log("Missing expected OPDS URL.")
      return nil
    }
  }

  @objc init(title: String, books: [Any]) {
    super.init()
    self.humanReadableDescription = title
    self.books = books
  }

  @objc func opdsURL(forSearchingString searchString: String) -> URL? {
    guard let template = opdsURLTemplate,
          let encoded = (searchString as NSString).stringURLEncodedAsQueryParamValue() else {
      return nil
    }
    let urlStr = template.replacingOccurrences(of: "{searchTerms}", with: encoded)
    return URL(string: urlStr)
  }
}
